tests/sync/app.cpp

```diff
@@ -1480,6 +1482,77 @@ TEST_CASE("app: remote mongo client", "[sync][app]") {
     }
 }
 
+TEST_CASE("app: push notifications", "[sync][app]") {
+    
+    std::unique_ptr<GenericNetworkTransport> (*factory)() = []{
+        return std::unique_ptr<GenericNetworkTransport>(new IntTestTransport);
+    };
+    std::string base_url = get_base_url();
+    std::string config_path = get_config_path();
+    REQUIRE(!base_url.empty());
+    REQUIRE(!config_path.empty());
+    auto config = App::Config {
+        get_runtime_app_id(config_path),
+        factory,
+        base_url,
+        util::none,
+        Optional<std::string>("A Local App Version"),
+        util::none,
+        "Object Store Platform Tests",
+        "Object Store Platform Version Blah",
+        "An sdk version"
+    };
+    
+    auto app = App::get_shared_app(config);
+    std::string base_path = tmp_dir() + "/" + config.app_id;
+    reset_test_directory(base_path);
+    TestSyncManager init_sync_manager(base_path);
+    
+    auto email = util::format("realm_tests_do_autoverify%1@%2.com", random_string(10), random_string(10));
+    auto password = random_string(10);
+    
+    app->provider_client<App::UsernamePasswordProviderClient>()
+    .register_email(email,
+                    password,
+                    [&](Optional<app::AppError> error) {
+                        CHECK(!error);
+                    });
+    
+    std::shared_ptr<SyncUser> sync_user;
+    
+    app->log_in_with_credentials(realm::app::AppCredentials::username_password(email, password),
+                                [&](std::shared_ptr<realm::SyncUser> user, Optional<app::AppError> error) {
+        REQUIRE(user);
+        CHECK(!error);
+        sync_user = user;
+    });
+    
+    SECTION("register") {
+        bool processed;
+        
+        app->push_notification_client("gcm").register_device("hello",
+                                                                  sync_user,
+                                                                  [&](Optional<app::AppError> error) {
+            CHECK(!error);
+            processed = true;
```

Comment:
could also check that the user here is the same as the one that we logged in with. Also would be nice to check some error conditions around this (and making sure the error messages are clear), such as:
- what happens when registering a device when there is no current user logged in?
- what happens when registering to a push service that doesn't exist?
- what happens when registering twice for the same user?
- what happens when deregistering for a user that hasn't registered?